File: docs/js/classes.js
import { CONFIG } from './config.js';

export class Camera {
    constructor() {
        this.x = 0;
        this.y = 0;
        this.shakeStrength = 0;
        this.shakeDamp = 0.9;
    }

    shake(amount) {
        this.shakeStrength = amount;
    }

    update() {
        if (this.shakeStrength > 0.5) {
            this.x = (Math.random() - 0.5) * this.shakeStrength;
            this.y = (Math.random() - 0.5) * this.shakeStrength;
            this.shakeStrength *= this.shakeDamp;
        } else {
            this.x = 0;
            this.y = 0;
            this.shakeStrength = 0;
        }
    }
}

export class SoundFX {
    constructor() {
        this.ctx = new (window.AudioContext || window.webkitAudioContext)();
        this.masterGain = this.ctx.createGain();
        this.masterGain.gain.value = 0.3;
        this.masterGain.connect(this.ctx.destination);
        this.initialized = false;
    }

    init() {
        if (!this.initialized) {
            this.ctx.resume().then(() => {
                this.initialized = true;
            });
        }
    }

    playTone(freq, type, duration, slideTo = null) {
        if (this.ctx.state === 'suspended') this.init();
        const osc = this.ctx.createOscillator();
        const gain = this.ctx.createGain();

        osc.type = type;
        osc.frequency.setValueAtTime(freq, this.ctx.currentTime);

        if (slideTo) {
            osc.frequency.exponentialRampToValueAtTime(slideTo, this.ctx.currentTime + duration);
        }

        gain.gain.setValueAtTime(0.3, this.ctx.currentTime);
        gain.gain.exponentialRampToValueAtTime(0.01, this.ctx.currentTime + duration);

        osc.connect(gain);
        gain.connect(this.masterGain);
        osc.start();
        osc.stop(this.ctx.currentTime + duration);
    }

    playNoise(duration) {
        if (this.ctx.state === 'suspended') this.init();
        const bufferSize = this.ctx.sampleRate * duration;
        const buffer = this.ctx.createBuffer(1, bufferSize, this.ctx.sampleRate);
        const data = buffer.getChannelData(0);

        for (let i = 0; i < bufferSize; i++) {
            data[i] = Math.random() * 2 - 1;
        }

        const noise = this.ctx.createBufferSource();
        noise.buffer = buffer;
        const gain = this.ctx.createGain();

        const filter = this.ctx.createBiquadFilter();
        filter.type = 'lowpass';
        filter.frequency.value = 1000;

        gain.gain.setValueAtTime(0.5, this.ctx.currentTime);
        gain.gain.exponentialRampToValueAtTime(0.01, this.ctx.currentTime + duration);

        noise.connect(filter);
        filter.connect(gain);
        gain.connect(this.masterGain);
        noise.start();
    }

    // --- PRESETS ---
    shoot() {
        this.playTone(800, 'square', 0.1, 100);
    } // "Pew"
    chargedShoot() {
        this.playTone(400, 'sawtooth', 0.4, 500);
    } // deep "Zap"
    charge() {
        this.playTone(200, 'sine', 0.1, 600);
    } // pitch up
    mineDrop() {
        this.playTone(600, 'sine', 0.1, 300);
    }
    explosion() {
        this.playNoise(0.4);
    } // "Kshhh"

    win() {
        [440, 554, 659, 880].forEach((f, i) => setTimeout(() => this.playTone(f, 'square', 0.2), i * 150));
    }

    powerup() {
        this.playTone(400, 'sine', 0.1);
        setTimeout(() => this.playTone(600, 'sine', 0.1), 100);
    }
    start() {
        this.playTone(440, 'triangle', 0.5, 880);
    }
    death() {
        const melody = [500, 400, 300, 200, 100, 50, 10];
        melody.forEach((freq, index) => {
            setTimeout(() => {
                this.playTone(freq, 'triangle', 0.1, freq - 50);
            }, index * 120);
        });
    }
    roundOver() {
        this.playTone(880, 'square', 0.15); // Beep
        setTimeout(() => this.playTone(440, 'square', 0.4), 200); // Boop (Longer)
    }
    shield() {
        this.playTone(100, 'sine', 0.25, 800);
    }
    niaNiaNia() {
        const melody = [880, 784, 880, 784, 880, 784, 880];
        melody.forEach((freq, index) => {
            setTimeout(() => {
                this.playTone(freq, 'triangle', 0.1, freq - 50);
            }, index * 120);
        });
    }
    boost() {
        // Start at 60Hz, slide to 100Hz, duration 0.1s
        // 'sawtooth' gives it a buzzy, mechanical feel

        this.playTone(60, 'sine', 0.9, 850);
    }
}

export class Cell {
    constructor(c, r) {
        this.c = c;
        this.r = r;
        this.walls = [true, true, true, true];
        this.visited = false;
        this.parent = null;
        this.bfsVisited = false;
    }
}

export class Player {
    constructor(id, color, controls) {
        this.id = id;
        this.name = `CPU`;
        this.color = color;
        this.controls = controls;
        this.size = 2.0;
        this.score = 0;
        this.goalC = 0;
        this.goalR = 0;
        this.lastDir = {
            x: id === 0 ? 1 : -1,
            y: 0
        };

        // AI Memory
        this.lastPos = {
            x: 0,
            y: 0
        };
        this.stuckCounter = 0;
        this.forceUnstuckTimer = 0;
        this.unstuckDir = {
            x: 0,
            y: 0
        };
        this.resetState();
    }

    resetState() {
        this.minesLeft = CONFIG.MAX_MINES;
        this.lastMineTime = 0;
        this.lastBoostTime = 0;
        this.trail = [];
        this.boostEnergy = 100;
        this.boostCooldown = 0;
        this.portalCooldown = 0;
        this.stunTime = 0;
        this.shieldActive = false;
        this.currentSpeed = CONFIG.BASE_SPEED;
        this.prevDetonateKey = false;
        this.beamPixels = [];
        this.beamIdx = 0;
        this.isCharging = false;
        this.chargeStartTime = 0;
        this.glitchTime = 0;
        this.chargeGrace = 0;
        this.botPath = [];
        this.botNextCell = null;
        this.botRetargetTimer = 0;
        this.stuckCounter = 0;
        this.forceUnstuckTimer = 0;
        this.isDead = false;
        this.ai = null;
    }
}

File: pi/classes.py
# pi/classes.py
# Game entity classes ported from docs/js/classes.js
# SoundFX is browser-only (Web Audio API) and is intentionally omitted.

import random as _random
from config import MAX_MINES, MAX_ENERGY, BASE_SPEED, TIMING


class Cell:
    """Maze cell — mirrors JS Cell class."""
    __slots__ = ('c', 'r', 'walls', 'visited', 'parent', 'bfs_visited', 'gCost')

    def __init__(self, c: int, r: int) -> None:
        self.c = c
        self.r = r
        self.walls = [True, True, True, True]  # [N, E, S, W]
        self.visited = False
        self.parent = None
        self.bfs_visited = False
        self.gCost = float('inf')  # A* cost — reset before each pathfinding call


class Camera:
    """Camera with shake effect — mirrors JS Camera class."""
    __slots__ = ('x', 'y', 'shake_strength', 'shake_damp')

    def __init__(self) -> None:
        self.x = 0.0
        self.y = 0.0
        self.shake_strength = 0.0
        self.shake_damp = 0.9

    def shake(self, amount: float) -> None:
        """Set shake strength (matches JS camera.shake(amount))."""
        self.shake_strength = amount

    def update(self) -> None:
        """Advance shake one frame — matches JS Camera.update()."""
        if self.shake_strength > 0.5:
            self.x = (_random.random() - 0.5) * self.shake_strength
            self.y = (_random.random() - 0.5) * self.shake_strength
            self.shake_strength *= self.shake_damp
        else:
            self.x = 0.0
            self.y = 0.0
            self.shake_strength = 0.0


class Player:
    """
    Player entity — mirrors JS Player class.

    Constructor arguments:
      id       -- integer player id (required)
      name     -- display name string (default 'CPU')
      color    -- color value / Color object (default None)
      controls -- input map dict (default None)
      x, y     -- initial position (default 0.0)
      size     -- hitbox size (default 2.0)
    """
    # Use __slots__ to keep attribute layout explicit and lean.
    __slots__ = (
        'id', 'name', 'color', 'controls',
        'size', 'score', 'goal_c', 'goal_r',
        'x', 'y',
        # Per-round state (reset by reset_state)
        'mines_left', 'last_mine_time', 'last_boost_time',
        'trail', 'boost_energy', 'boost_cooldown',
        'portal_cooldown', 'portal_invuln_frames',
        'shield_active', 'current_speed', 'prev_detonate_key',
        'beam_pixels', 'beam_idx',
        'is_charging', 'charge_start_time',
        'glitch_start_time', 'stun_start_time',
        'bot_path', 'bot_next_cell', 'bot_retarget_timer',
        'force_unstuck_timer', 'stuck_counter',
        'is_dead', 'ai',
        # AI memory
        'last_pos', 'last_dir', 'unstuck_dir',
        'ai_mental_model', 'ai_frame_counter',
        'confusion_timer', 'confused_dir',
        'direction_history', '_suggested_mine_pos',
    )

    def __init__(
        self,
        id: int,
        name: str = 'CPU',
        color=None,
        controls=None,
        x: float = 0.0,
        y: float = 0.0,
        size: float = 2.0,
    ) -> None:
        self.id = id
        self.name = name
        self.color = color
        self.controls = controls
        self.size = size
        self.score = 0
        self.goal_c = 0
        self.goal_r = 0
        self.x = x
        self.y = y
        self.reset_state()

    def reset_state(self) -> None:
        """
        Reset all per-round state.  Score is intentionally NOT reset.
        Mirrors JS Player.resetState().
        """
        self.mines_left = MAX_MINES
        self.last_mine_time = 0
        self.last_boost_time = 0
        self.trail = []
        self.boost_energy = MAX_ENERGY
        self.boost_cooldown = 0
        self.portal_cooldown = 0
        self.portal_invuln_frames = 0
        self.shield_active = False
        self.current_speed = BASE_SPEED
        self.prev_detonate_key = False
        self.beam_pixels = []
        self.beam_idx = 0
        self.is_charging = False
        self.charge_start_time = None  # None = not currently charging
        self.glitch_start_time = 0
        self.stun_start_time = 0
        self.bot_path = []
        self.bot_next_cell = None
        self.bot_retarget_timer = 0
        self.force_unstuck_timer = 0
        self.stuck_counter = 0
        self.is_dead = False
        self.ai = None
        # AI memory
        self.last_pos = {'x': 0.0, 'y': 0.0}
        self.last_dir = {'x': 0.0, 'y': 0.0}
        self.unstuck_dir = {'x': 0.0, 'y': 0.0}
        self.ai_mental_model = None
        self.ai_frame_counter = 0
        self.confusion_timer = 0
        self.confused_dir = None
        self.direction_history = []
        self._suggested_mine_pos = None

    # ------------------------------------------------------------------
    # Timing helpers — all mirror the JS equivalents
    # ------------------------------------------------------------------

    def glitch_remaining(self, frame: int) -> int:
        """Frames remaining in current glitch effect."""
        time_diff = frame - self.glitch_start_time
        return TIMING.GLITCH_DURATION - time_diff

    def glitch_is_active(self, frame: int) -> bool:
        """True if player is currently glitching."""
        if self.glitch_start_time != 0:
            time_diff = frame - self.glitch_start_time
            return time_diff < TIMING.GLITCH_DURATION
        return False

    def stun_remaining(self, frame: int) -> int:
        """Frames remaining in current stun."""
        time_diff = frame - self.stun_start_time
        return TIMING.STUN_DURATION - time_diff

    def stun_is_active(self, frame: int) -> bool:
        """True if player is currently stunned."""
        if self.stun_start_time != 0:
            time_diff = frame - self.stun_start_time
            return time_diff < TIMING.STUN_DURATION
        return False

    def charge_is_ready(self, frame: int) -> bool:
        """True if charged beam has been held for CHARGE_DURATION frames.

        Note: charge_start_time=None means charging hasn't begun.
        When charge_start_time is an integer (including 0), the elapsed
        frame count is evaluated arithmetically, matching the test spec.
        """
        if self.charge_start_time is not None:
            time_diff = frame - self.charge_start_time
            return time_diff >= TIMING.CHARGE_DURATION
        return False
